### PR TITLE
 Add keyboard shortcut for changing the board #919 

### LINKED ISSUE #919: Add keyboard shortcut for changing the board
Suggestion: Ctrl+B

### PR DESCRIPTION
**Pull request by HyungJon**
Fixes #919 

Changes board when Ctrl+B is pushed.
Only happens if 1) there are at least 2 boards saved, 2) one of them is currently open

2 might be debatable. I chose that because I wanted to prevent changing the board and losing the current board panels. I'll modify if there are other opinions.

### CODE REVIEW

**Comment by damithc** [comment]
Showing the current board name should be done soon too. Without it, this feature is a big hard to use.

File: src/test/java/guitests/MenuControlTest.java [hunk 1]
```diff
@@ -21,6 +22,8 @@ public void menuControlTest() {
         modelUpdatedEventTriggered = false;
         UI.events.registerEvent((ModelUpdatedEventHandler) e -> modelUpdatedEventTriggered = true);
         PanelControl panelControl = (PanelControl) find("#dummy/dummy_col0").getParent();
+        Preferences testPref = panelControl.getPreferences();
```
Comment (dariusf) [inline src/test/java/guitests/MenuControlTest.java]: Can we use `UI.prefs` instead?

File: src/main/java/ui/issuepanel/PanelControl.java [hunk 2]
```diff
@@ -273,4 +273,8 @@ public int getNumberOfPanels() {
     public int getNumberOfSavedBoards() {
         return prefs.getAllBoards().size();
     }
+    
+    public Preferences getPreferences() {
```
Comment (dariusf) [inline src/main/java/ui/issuepanel/PanelControl.java]: We wouldn't need this getter in that case.

File: src/main/java/ui/MenuControl.java [hunk 3]
```diff
@@ -271,6 +272,20 @@ private void onBoardDelete(String boardName) {
             }
         }).collect(Collectors.toList());
     }
+    
+    public void switchBoard() {
+        if (prefs.getLastOpenBoard().isPresent() && prefs.getAllBoards().size() > 1) {
+            List<String> boardNames = new ArrayList<>(prefs.getAllBoards().keySet());
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: This is backed by a `HashMap`, which doesn't guarantee key ordering: the boards may be cycled in a different order [over time](http://stackoverflow.com/a/2144822). You might want to change it to a `LinkedHashMap`.

File: src/main/java/ui/MenuControl.java [hunk 4]
```diff
@@ -271,6 +272,20 @@ private void onBoardDelete(String boardName) {
...
+            int index = boardNames.indexOf(prefs.getLastOpenBoard().get());
+            if (index == boardNames.size() - 1) {
+                index = 0;
+            } else {
+                index++;
+            }
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: Use `%` instead?
```java
int index = (lastBoard + 1) % boardNames.size();
```

**Comment by dariusf** [comment]
Could you also see if the `sleep`s in tests can be replaced with `PlatformEx.waitOnFxThread`?

File: src/test/java/guitests/MenuControlTest.java (hunk 1 above)
Comment (HyungJon) [inline src/test/java/guitests/MenuControlTest.java]: That will result in non-static reference error. Can I move the ```getPrefs()``` from ```PanelControl``` to ```UI``` for this purpose?

File: src/main/java/ui/MenuControl.java (hunk 3 above)
Comment (HyungJon) [inline src/main/java/ui/MenuControl.java]: Do I just have to change ```savedBoards``` in GlobalConfig to LinkedHashMap, or should every access to it be changed accordingly to ensure the order will be retained?

**Comment by HyungJon** [comment]
Confirmed that ```MenuControlTest``` runs okay with ```PlatformEx.waitOnFxThread```. Shall I create a new issue for changing all ```sleep``` with ```PlatformEx.waitOnFxThread```? #739 has a related discussion though.

File: src/test/java/guitests/MenuControlTest.java (hunk 1 above)
Comment (dariusf) [inline src/test/java/guitests/MenuControlTest.java]: Right, `prefs` isn't static. Moving it to `UI` would still require a reference to the current `UI` instance though. How about making `prefs` static, then doing that?

File: src/main/java/ui/MenuControl.java (hunk 3 above)
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: Just the former should do it. What do you mean by changing every access?

It'd be good to add a non-GUI test for this too. Save and delete a few boards, then test that cycling returns stuff in the expected order.

**Comment by dariusf** [comment]
Great! We should get rid of `sleep` eventually, but I'm not sure yet if in general it's possible to mechanically replace it with `waitOnFxThread`. I guess we should do it test by test for now.

That issue can be repurposed. `waitOnFxThread` is generally what we want when we use `sleep` in tests, not `runAndWait`.

File: src/test/java/guitests/MenuControlTest.java (hunk 1 above)
Comment (dariusf) [inline src/test/java/guitests/MenuControlTest.java]: You can exclude it [here](https://github.com/HubTurbo/HubTurbo/blob/master/config/findbugs/excludeFilter.xml#L23-L30).

**Comment by HyungJon** [comment]
Oops, forgot to commit the last fix to findbugsTest. Squashed and pushed.

I'm really unable to figure out what this status check thing is. Looks like I should be able to fix and push it from my end, how can I do it?

**Comment by dariusf** [comment]
Status check? You mean the drop in coverage?

**Comment by HyungJon** [comment]
The branch became unmergeable because "Required status must pass before merging". Probably because I edited the excludeFilter I think?

**Comment by dariusf** [comment]
That would cause the Travis build to fail if it were an issue. The only remaining check I see is the coveralls one. How were you trying to merge?

**Comment by damithc** [comment]
Merging should be done locally, not using GitHub.

**Comment by HyungJon** [comment]
cap 2015-09-05 10-59-08-132

This is displayed. It's visible only to the branch owner then?

**Comment by HyungJon** [comment]
Oh wait, just branch out of date I think.

**Comment by dariusf** [comment]
That's a new GitHub feature that enforces that all pre-merge checks have passed before allowing a merge.

Isn't the problem that coverage has dropped?

**Comment by HyungJon** [comment]
Ah, it makes sense now. I think the coverage decrease might be because I added a simplified version of ```switchBoard``` to the non-GUI test as well. I think I should totally move ```switchBoard``` implementation to ```Preferences``` and make ```MenuControl``` only call that function.

**Comment by HyungJon** [comment]
Bingo. But I'm really really not a fan of the change I had to make (calling ```switchBoard``` normally with Ctrl+B will result in one redundant ```setLastOpenBoard```), so if there's any way to achieve higher coverage without doing this, I'll immediately do that.

**Comment by dariusf** [comment]
I think it's fine. Can't think of a clearer way to write it.

**Pull request merged**